dimod/generators/wireless.py

```diff
@@ -0,0 +1,985 @@
+# Copyright 2023 D-Wave Systems Inc.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express 93or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+from collections import namedtuple
+from functools import wraps
+from itertools import product
+from typing import Literal, Optional, Tuple, Union
+
+import numpy as np
+
+import dimod
+
+__all__ = ['mimo', 'comp', ]
+
+mod_params = namedtuple("mod_params", ["bits_per_transmitter", 
+                                       "amplitudes", 
+                                       "transmitters_per_spin", 
+                                       "number_of_amps", 
+                                       "spins_per_symbol"])
+mod_config = {
+    "BPSK": mod_params(1, 1, 1, 1, 1),
+    "QPSK": mod_params(2, 1, 2, 1, 1),
+    "16QAM": mod_params(4, 2, 4, 2, 2),
+    "64QAM": mod_params(6, 4, 6, 3, 3),
+    "256QAM": mod_params(8, 8, 8, 5, 4)}
+
+def _quadratic_form(y, F):
+    """Convert :math:`O(v) = ||y - F v||^2` to sparse quadratic form.
+
+    Constructs coefficients for the form
+    :math:`O(v) = v^{\dagger} J v - 2 \Re(h^{\dagger} v) + \\text{offset}`.
+
+    Args:
+        y: Received signal as a NumPy column vector of complex or real values.
+
+        F: Wireless channel as an :math:`i \\times j` NumPy matrix of complex
+            values, where :math:`i` rows correspond to :math:`y_i` receivers
+            and :math:`j` columns correspond to :math:`v_i` transmitted symbols.
+
+    Returns:
+        Three tuple of offset, as a real scalar, linear biases :math:`h`, as a dense
+        real vector, and quadratic interactions, :math:`J`, as a dense real symmetric
+        matrix.
+    """
+
+    if len(F.shape) != 2 or F.shape[0] != y.shape[0]:
+        raise ValueError("F should have shape (n, m) for some m, n "
+                         "and n should equal y.shape[1];"
+                         f" given: {F.shape}, n={y.shape[1]}")
+
+    offset = np.matmul(y.imag.T, y.imag) + np.matmul(y.real.T, y.real)
+    h = -2 * np.matmul(F.T.conj(), y)    # Be careful with interpretation!
```

Comment:
Same comment as before, it's not at all obvious how I should be careful here. This comment should be expanded or removed.

Reply:
Remove it

Reply:
Removed